components/ranker/__init__.py

```diff
@@ -0,0 +1,305 @@
+        and the list items are numpy arrays of float for each time slot during the specified night.
+        """
+        # Scores are indexed by night_idx and contain scores for each time slot.
+        # We initialize to all zeros.
+        scores = deepcopy(self._zero_scores[obs.site])
+
+        # target_info is a map from night index to TargetInfo.
+        # We require it to proceed for hour angle / elevation information and coordinates.
+        # If it is missing, just return scores of 0.
+        target_info = Collector.get_target_info(obs.id)
+        if target_info is None:
+            return scores
+
+        # TODO: Check this for correctness in mapping from old calcs to new.
+        remaining = obs.exec_time() - obs.total_used()
+        cplt = (program.total_used() + remaining) / program.total_awarded()
+
+        metric, metric_s = self._metric_slope(np.array([cplt]),
+                                              np.ones(1, dtype=int) * program.band,
+                                              np.ones(1) * 0.8,
+                                              program.thesis)
+
+        # Declination for the base target per night.
+        dec = [target_info[night_idx].coord.dec for night_idx in self.night_indices]
+
+        # Hour angle / airmass
+        ha = [target_info[night_idx].hourangle for night_idx in self.night_indices]
+
+        # Get the latitude associated with the site.
+        site_latitude = obs.site.value.location.lat
+        if site_latitude < 0. * u.deg:
+            dec_diff = [site_latitude - np.max(dec[night_idx]) for night_idx in self.night_indices]
+        else:
+            dec_diff = [np.min(dec[night_idx]) - site_latitude for night_idx in self.night_indices]
+
+        if np.abs(dec_diff) < 40. * u.deg:
+            # Weighted to slightly positive HA.
+            c = np.array([3., 0.1, -0.06])
+        else:
+            # Weighted to 0 HA if Xmin > 1.3.
+            c = np.array([3., 0., -0.08])
+
+        wha = [c[0] + c[1] * ha[night_idx] / u.hourangle + (c[2] / u.hourangle ** 2) * ha[night_idx] ** 2
+               for night_idx in self.night_indices]
+        kk = [np.where(wha[night_idx] <= 0.)[0] for night_idx in self.night_indices]
+        for night_idx in self.night_indices:
+            wha[night_idx][kk[night_idx]] = 0.
+
+        p = [(metric[0] ** self.params.met_power) *
+             (target_info[night_idx].rem_visibility_frac ** self.params.vis_power) *
+             (wha[night_idx] ** self.params.wha_power)
+             for night_idx in self.night_indices]
+
+        # Assign scores in p to all indices where visibility constraints are met.
+        # They will otherwise be 0 as originally defined.
+        for night_idx in self.night_indices:
+            slot_indices = target_info[night_idx].visibility_slot_idx
+            scores[night_idx][slot_indices] = p[slot_indices]
+
+        return scores
+
+    def get_observation_scores(self, obs_id: ObservationID) -> Scores:
```

Comment:
Bryan and I talked about this and `Observation` can only be associated with a single site... if we have an `Observation` that can be executed at more than one site, we'd create a second `Observation` to represent that and then link them with an `ORGroup` as you suggest here.

It will take more effort, but since there are so many site-specific elements involved in the scoring, I'm not sure that there's much we can do to get around it.